build.gradle.kts

```diff
@@ -0,0 +1,268 @@
+        }
+    }
+}
+
+apply(plugin = "gradlebuild.cleanup")
+apply(plugin = "gradlebuild.available-java-installations")
+apply(plugin = "gradlebuild.buildscan")
+apply(from = "gradle/versioning.gradle")
+apply(from = "gradle/dependencies.gradle")
+apply(plugin = "gradlebuild.minify")
+apply(from = "gradle/testDependencies.gradle")
+apply(plugin = "gradlebuild.wrapper")
+apply(plugin = "gradlebuild.ide")
+apply(plugin = "gradlebuild.no-resolution-at-configuration")
+apply(plugin = "gradlebuild.update-versions")
+apply(plugin = "gradlebuild.dependency-vulnerabilities")
+apply(plugin = "gradlebuild.add-verify-production-environment-task")
+
+
+subprojects {
+    version = rootProject.version
+
+    if (project in javaProjects) {
+        apply(plugin = "gradlebuild.java-projects")
+        apply(plugin = "gradlebuild.dependencies-metadata-rules")
+    }
+
+    if (project in publishedProjects) {
+        apply(plugin = "gradlebuild.publish-public-libraries")
+    }
+
+    apply(from = "$rootDir/gradle/shared-with-buildSrc/code-quality-configuration.gradle.kts")
+    apply(plugin = "gradlebuild.task-properties-validation")
+    apply(plugin = "gradlebuild.test-files-cleanup")
+}
+
+configurations {
+    "coreRuntime" {
+        attributes { attribute(Usage.USAGE_ATTRIBUTE, project.objects.named(Usage::class.java, Usage.JAVA_RUNTIME)) }
+        isCanBeResolved = true
+        isCanBeConsumed = false
+        isVisible = false
+    }
+    "coreRuntimeExtensions" {
+        attributes { attribute(Usage.USAGE_ATTRIBUTE, project.objects.named(Usage::class.java, Usage.JAVA_RUNTIME)) }
+        isCanBeResolved = true
+        isCanBeConsumed = false
+        isVisible = false
+    }
+    "externalModules" {
+        isVisible = false
+    }
+    // Configuration used to resolve external modules before patching them with versions from core runtime
+    "externalModulesRuntime" {
+        isVisible = false
+        extendsFrom(configurations["coreRuntime"])
+        extendsFrom(configurations["externalModules"])
+    }
+    // Combines the 'coreRuntime' with the patched external module jars
+    "runtime" {
+        isVisible = false
+        extendsFrom(configurations["coreRuntime"])
+    }
+    "gradlePlugins" {
+        isVisible = false
+    }
+    "testRuntime" {
+        extendsFrom(configurations["runtime"])
+        extendsFrom(configurations["gradlePlugins"])
+    }
+
+    all {
+        attributes.attribute(Usage.USAGE_ATTRIBUTE, objects.named(Usage::class.java, Usage.JAVA_RUNTIME))
+    }
+}
+
+extra.set("allTestRuntimeDependencies", configurations["testRuntime"].allDependencies)
+
+val patchedExternalModulesDir = File(buildDir, "external/files")
+val patchedExternalModules = files(Callable { fileTree(patchedExternalModulesDir).files.sorted() })
+patchedExternalModules.builtBy("patchExternalModules")
+
+dependencies {
+    "externalModules"("org.gradle:gradle-kotlin-dsl:${BuildEnvironment.gradleKotlinDslVersion}")
```

Comment:
:o: consider bringing all the `Configuration`s into scope when creating them so there's no need to repeat strings to reference them here.

e.g.
```kotlin
val externalModules by configurations.creating {
    visible = false
}
// ...
dependencies {
    externalModules("...")
}
```